Fix stack alignment for JIT function call targets

DresslerFrank opened the pull request:
This change fixes a SIGSEGV error caused by a violation of the X84_64 ABI which demands that the stack pointer is aligned to a 16-byte boundary prior to a CALL. The commit message contains a bit more info.

This probably resolves issue #128.

Another variant of fixing this issue is implemented in https://github.com/DresslerFrank/dynamips/commit/0db6692d11a820e92ee6b7a6b25ad2a751508853

spikefishjohn commented:
@DresslerFrank oooh nice! Thank you so much for this. I'll poke this tomorrow. 

If you don't mind, how did you figure that out? I went on a epic journey to learn some GDB without only a single asm class from.. uh.. 15 years ago. Needless to say I didn't get far, but it was fascinating. 

Is this common to Ubuntu 22.04?

grossmj commented:
Thanks a lot for this. I am going to rebuild the Debian package for 22.04 soon.

DresslerFrank commented:
@grossmj Thank you for the work.

DresslerFrank commented:
### Reproducing the Bug
1. Checkout 61a198d4
2. Build
```
   mkdir build
   cd build
   cmake ..
   nice -n15 make -j$(nproc)
```
3. Run:
  ./stable/dynamips -r 1024 -X -P 1700 <some 1700 .image>
  
Result:
```
...
Segmentation fault (core dumped)
```

### Locating the Bug
1. Rebuild with debug info
```
   cmake -DCMAKE_BUILD_TYPE=Debug ..
   nice -n15 make -j$(nproc)
```
2. Run in gdb:
```
gdb -q --args ./stable/dynamips -r 1024 -X -P 1700 <some 1700 .image>
(gdb) run
...
Thread 15 "dynamips" received signal SIGSEGV, Segmentation fault.
...
(gdb) bt
#0  0x00007ffff7debb6e in tzset_internal (always=0) at ./time/tzset.c:372
...
#10 ppc32_stw (cpu=0x55555567f010, vaddr=4127195196, reg=<optimized out>) at .../dynamips/stable/ppc32_mem.c:808
#11 0x00007ffff0c9a4f4 in ?? ()
#12 0x000055555559aa90 in ppc32_jit_tcb_exec (block=<optimized out>, cpu=0x55555567f010) at .../dynamips/common/ppc32_amd64_trans.h:65
...
```

We see where it crashed (`#1`) and that it is probably running code generated by the JIT (function name in `#12`; and in `#11` the JIT code has no name, hence `??`).

Look at what line was executed:
```
(gdb) list
367	in ./time/tzset.c
```

Well, that is not enough info. We need to get the source code of "./time/tzset.c".

```
(gdb) info symbol tzset_internal 
tzset_internal in section .text of /lib/x86_64-linux-gnu/libc.so.6
```

So it is part of libc. "libc.so.6" is in package libc6.
```
$ apt-file search libc.so.6
libc6: /lib/x86_64-linux-gnu/libc.so.6  
...
```

So we get its source code with

```
apt source libc6
```

and look for the `tzset.c`

```
$ find -name tzset.c
./glibc-2.35/time/tzset.c

```

Good, we have found it. We tell gdb do load the code.

```
(gdb) dir .../glibc-2.35
```

And try to list again:
```
(gdb) list ./time/tzset.c:372
367	tzset_internal (int always)
368	{
369	  static int is_initialized;
370	  const char *tz;
371	
372	  if (is_initialized && !always)
373	    return;
374	  is_initialized = 1;
375	
376	  /* Examine the TZ environment variable.  */
```

Line 372 where the SIGSEGV happened according to GDB does not look like it could cause a SIGSEGV. But the machine executes machine code, not C, so there might be an instruction that GDB cannot map to a meaningful line. So let's ignore the C listing and look at the machine code.

```
(gdb) disas
Dump of assembler code for function tzset_internal:
   0x00007ffff7debb50 <+0>:	push   %r12
   0x00007ffff7debb52 <+2>:	push   %rbp
   0x00007ffff7debb53 <+3>:	push   %rbx
   0x00007ffff7debb54 <+4>:	lea    0xfefc0(%rip),%rbx        # 0x7ffff7eeab1b
   0x00007ffff7debb5b <+11>:	movq   %rbx,%xmm0
   0x00007ffff7debb60 <+16>:	punpcklqdq %xmm0,%xmm0
   0x00007ffff7debb64 <+20>:	sub    $0x10,%rsp
   0x00007ffff7debb68 <+24>:	mov    0x145b7e(%rip),%ecx        # 0x7ffff7f316ec <is_initialized.0>
=> 0x00007ffff7debb6e <+30>:	movaps %xmm0,(%rsp)
...
```

How can this instruction cause a SIGSEGV? `%rsp` might point somewhere that we must not write. So get `%rsp` and the process ID:

```
(gdb) info proc
process 16094
...
(gdb) p $rsp
$1 = (void *) 0x7fffec8ac9b8
```

and check for the access permissions for the page of process `16094` that contains the address `0x7fffec8ac9b8` (`%rsp`).

$ sudo ./addr_perm.sh 16094 0x7fffec8ac9b8
7fffec0ae000-7fffec8ae000 rw-p 00000000 00:00 0

So it should it writeable.

The script `addr_perm.sh` is here:
```sh
#!/usr/bin/env bash

[[ $# -ne 2 ]] && echo "Usage: $0 <pid> <address>" && exit 1

PID="$1"
ADDRESS="$2"

gawk -v address="$ADDRESS" -f - "/proc/$PID/maps" <<'EOF'
BEGIN { 
        address_dec = strtonum(address)
}
match($0, /([^-]*)-([^ ]*)/, captures) {
        a = strtonum("0x" captures[1])
        o = strtonum("0x" captures[2])
        if (address_dec < a) {
                next
        }
        if (address_dec > o) {
                next
        }
        print
}
EOF
```

So what can be the reason that `MOVAPS` causes a SIGSEGV? Google: "MOVAPS sigsegv"
-> https://stackoverflow.com/questions/67243284/why-movaps-causes-segmentation-fault

We could check whether the answer is correct, but we do not need to do that because the name "Peter Cordes" is under the answer. His name indicates quality.

The memory addresses used with `MOVAPS` need to be aligned to a 16-byte boundary. The address from above, `0x7fffec8ac9b8`, is not. So that is the problem.

There is `MOVUPS` for unaligned accesses, but we cannot change libc. So that is not an option.

How can we get `RSP` aligned? Is it always misaligned?

I want to check how often `tzset_internal` is called and how the callstack and `RSP` look.

The following GDB breakpoint runs some commands, to print the backtrace, `RSP` and then continue automatically.

```Fixing the Bug
(gdb) br tzset_internal 
Breakpoint 3 at 0x7ffff7debb50: file ./time/tzset.c, line 368.
(gdb) commands
Type commands for breakpoint(s) 3, one per line.
End with a line saying just "end".
>bt
>p $rsp
>cont
>end
(gdb) run
...
```

This results in
```
Thread 1 "dynamips" hit Breakpoint 3, tzset_internal (always=0) at ./time/tzset.c:368
368	{
#0  tzset_internal (always=0) at ./time/tzset.c:368
#1  0x00007ffff7dec617 in __tz_convert (timer=1653335222, use_localtime=use_localtime@entry=0, tp=tp@entry=0x7fffffffdf20) at ./time/tzset.c:577
#2  0x00007ffff7de9f91 in __GI___gmtime_r (t=t@entry=0x7fffffffdf10, tp=tp@entry=0x7fffffffdf20) at ./time/gmtime.c:26
#3  0x0000555555577cd2 in m_flog (fd=0x5555556619b0, module=module@entry=0x55555561288d "C1700_BOOT", fmt=fmt@entry=0x5555556123e8 "starting instance (CPU0 PC=0x%8.8x,idle_pc=0x%8.8x,JIT %s)\n", ap=ap@entry=0x7fffffffe0b0) at .../dynamips/common/utils.c:266
#4  0x000055555557836e in m_flog (fd=<optimized out>, module=module@entry=0x55555561288d "C1700_BOOT", fmt=fmt@entry=0x5555556123e8 "starting instance (CPU0 PC=0x%8.8x,idle_pc=0x%8.8x,JIT %s)\n", ap=ap@entry=0x7fffffffe0b0) at .../dynamips/common/utils.c:264
#5  0x0000555555585656 in vm_flog (ap=0x7fffffffe0b0, format=0x5555556123e8 "starting instance (CPU0 PC=0x%8.8x,idle_pc=0x%8.8x,JIT %s)\n", module=0x55555561288d "C1700_BOOT", vm=<optimized out>) at .../dynamips/stable/vm.c:260
#6  vm_log (vm=vm@entry=0x55555565f730, module=module@entry=0x55555561288d "C1700_BOOT", format=format@entry=0x5555556123e8 "starting instance (CPU0 PC=0x%8.8x,idle_pc=0x%8.8x,JIT %s)\n") at .../dynamips/stable/vm.c:270
#7  0x00005555555cd658 in c1700_boot_ios (router=<optimized out>, router=<optimized out>) at .../dynamips/common/dev_c1700.c:637
#8  c1700_init_instance (vm=<optimized out>) at .../dynamips/common/dev_c1700.c:687
#9  0x0000555555574b73 in main (argc=<optimized out>, argv=0x7fffffffe3c8) at .../dynamips/common/dynamips.c:958
$5 = (void *) 0x7fffffffdeb8

Thread 1 "dynamips" hit Breakpoint 3, tzset_internal (always=0) at ./time/tzset.c:368
368	{
#0  tzset_internal (always=0) at ./time/tzset.c:368
#1  0x00007ffff7dec617 in __tz_convert (timer=1653335222, use_localtime=use_localtime@entry=0, tp=tp@entry=0x7fffffffde00) at ./time/tzset.c:577
#2  0x00007ffff7de9f91 in __GI___gmtime_r (t=t@entry=0x7fffffffddf0, tp=tp@entry=0x7fffffffde00) at ./time/gmtime.c:26
#3  0x0000555555577cd2 in m_flog (fd=0x5555556619b0, module=module@entry=0x7fffffffdfb0 "CPU0: CPU_STATE", fmt=fmt@entry=0x555555608400 "Starting CPU (old state=%u)...\n", ap=ap@entry=0x7fffffffdf98) at .../dynamips/common/utils.c:266
#4  0x000055555557836e in m_flog (fd=<optimized out>, module=module@entry=0x7fffffffdfb0 "CPU0: CPU_STATE", fmt=fmt@entry=0x555555608400 "Starting CPU (old state=%u)...\n", ap=ap@entry=0x7fffffffdf98) at .../dynamips/common/utils.c:264
#5  0x00005555555855a5 in vm_flog (vm=<optimized out>, module=module@entry=0x7fffffffdfb0 "CPU0: CPU_STATE", format=format@entry=0x555555608400 "Starting CPU (old state=%u)...\n", ap=ap@entry=0x7fffffffdf98) at .../dynamips/stable/vm.c:260
#6  0x000055555558748f in cpu_log (cpu=cpu@entry=0x55555567ee00, module=module@entry=0x555555608448 "CPU_STATE", format=format@entry=0x555555608400 "Starting CPU (old state=%u)...\n") at .../dynamips/stable/cpu.c:181
#7  0x00005555555877e5 in cpu_start (cpu=0x55555567ee00) at .../dynamips/stable/cpu.c:273
#8  cpu_start (cpu=0x55555567ee00) at .../dynamips/stable/cpu.c:270
#9  0x00005555555cd6b6 in c1700_boot_ios (router=<optimized out>, router=<optimized out>) at .../dynamips/common/dev_c1700.c:644
#10 c1700_init_instance (vm=<optimized out>) at .../dynamips/common/dev_c1700.c:687
#11 0x0000555555574b73 in main (argc=<optimized out>, argv=0x7fffffffe3c8) at .../dynamips/common/dynamips.c:958
$6 = (void *) 0x7fffffffdd98
ROMMON emulation microcode.

[Switching to Thread 0x7fffeccb5640 (LWP 19558)]

Thread 15 "dynamips" hit Breakpoint 3, tzset_internal (always=0) at ./time/tzset.c:368
368	{
#0  tzset_internal (always=0) at ./time/tzset.c:368
#1  0x00007ffff7dec617 in __tz_convert (timer=1653335222, use_localtime=use_localtime@entry=0, tp=tp@entry=0x7fffeccb4a48) at ./time/tzset.c:577
#2  0x00007ffff7de9f91 in __GI___gmtime_r (t=t@entry=0x7fffeccb4a38, tp=tp@entry=0x7fffeccb4a48) at ./time/gmtime.c:26
#3  0x0000555555577cd2 in m_flog (fd=0x5555556619b0, module=module@entry=0x555555610502 "ROM", fmt=fmt@entry=0x555555609748 "%s", ap=ap@entry=0x7fffeccb4bd8) at .../dynamips/common/utils.c:266
#4  0x000055555557836e in m_flog (fd=<optimized out>, module=module@entry=0x555555610502 "ROM", fmt=fmt@entry=0x555555609748 "%s", ap=ap@entry=0x7fffeccb4bd8) at .../dynamips/common/utils.c:264
#5  0x0000555555585656 in vm_flog (ap=0x7fffeccb4bd8, format=0x555555609748 "%s", module=0x555555610502 "ROM", vm=<optimized out>) at .../dynamips/stable/vm.c:260
#6  vm_log (vm=vm@entry=0x55555565f730, module=module@entry=0x555555610502 "ROM", format=format@entry=0x555555609748 "%s") at .../dynamips/stable/vm.c:270
#7  0x00005555555ab8bb in dev_remote_control_access (cpu=0x55555567ee00, dev=<optimized out>, offset=<optimized out>, op_size=<optimized out>, op_type=<optimized out>, data=<optimized out>) at .../dynamips/common/dev_remote.c:184
#8  0x0000555555597b4a in dev_access_fast (data=0x7fffeccb4cf0, op_type=1, op_size=4, offset=<optimized out>, dev_id=<optimized out>, cpu=<optimized out>) at .../dynamips/common/device.h:94
#9  ppc32_mem_access (data=0x7fffeccb4cf0, op_type=1, op_size=4, op_code=8, cid=1, vaddr=4127195196, cpu=0x55555567f010) at .../dynamips/stable/ppc32_mem.c:419
#10 ppc32_stw (cpu=0x55555567f010, vaddr=4127195196, reg=<optimized out>) at .../dynamips/stable/ppc32_mem.c:808
#11 0x00007ffff0c9a4f4 in ?? ()
#12 0x000055555559aa90 in ppc32_jit_tcb_exec (block=<optimized out>, cpu=0x55555567f010) at .../dynamips/common/ppc32_amd64_trans.h:65
#13 ppc32_jit_tcb_run (block=<optimized out>, cpu=0x55555567f010) at .../dynamips/stable/ppc32_jit.c:1221
#14 ppc32_jit_run_cpu (gen=<optimized out>) at .../dynamips/stable/ppc32_jit.c:1312
#15 0x00007ffff7da5b43 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#16 0x00007ffff7e37a00 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
$7 = (void *) 0x7fffeccb49e0

Thread 15 "dynamips" received signal SIGSEGV, Segmentation fault.
0x00007ffff7debb6e in tzset_internal (always=0) at ./time/tzset.c:372
372	  if (is_initialized && !always)
```

The first time two times it is called from `Thread 1` and `RSP` is `0x...8` (see `$5`, `$6`) which means it was divisible by 16 just before the `CALL` as required by the ABI. (the `CALL` subtracted 8 to `RSP`, so it was ...0 before the `CALL`) 

The third call is from `Thread 15` but `RSP` is `0x...0? (see `$7`), which violates the ABI.

So the question is if it was always violating the ABI or if this is just the case in Ubuntu 22.04.

In Ubuntu 18.04 I observed that dynamips was not crashing. So let's start a docker container.

```
$ docker run -v .../images/:/images -v .../dynamips:/dynamips --rm -it ubuntu:18.04
Unable to find image 'ubuntu:18.04' locally
18.04: Pulling from library/ubuntu
40dd5be53814: Pull complete 
Digest: sha256:d21b6ba9e19feffa328cb3864316e6918e30acfd55e285b5d3df1d8ca3c7fd3f
Status: Downloaded newer image for ubuntu:18.04
root@aeb881b2aa5b:/#
```

and install everything necessary there to build and debug `dynamips`

```
root@aeb881b2aa5b:/# apt update
...
root@aeb881b2aa5b:/# apt install gdb gcc libelf-dev cmake
...
root@aeb881b2aa5b:/# cd /dynamips
root@aeb881b2aa5b:/# mkdir build
root@aeb881b2aa5b:/# cd build
root@aeb881b2aa5b:/# cmake -DCMAKE_BUILD_TYPE=Debug ..
root@aeb881b2aa5b:/# nice -n15 make -j$(nproc)
...
```

and run it in `gdb` with a breakpoint that only stops when the interesting situation happens, i.e. when `RSP` is divisible by 16.

```
root@aeb881b2aa5b:/# gdb -q --args ./stable/dynamips -r 1024 -X -P 1700 /images/<some 1700 .image>
(gdb) br tzset_internal if $rsp % 16 == 0
(gdb) run
...
Thread 1 "dynamips" hit Breakpoint 3, __tz_convert (timer=0x7ffd3abbf040, use_localtime=0, tp=0x7ffd3abbf050) at tzset.c:584
...
```
And it stops. So even in Ubuntu 18.04 we had misaligned call stacks with `tzset_internal`. But it never crashes, no matter how often I `continue`. And executing `disas` also does not show any `MOVAPS` or similar instruction that would require alignment in `tzset_internal`.

So my guess was that at some point in the last years Ubuntu or libc was changed to allow newer instructions. Did they maybe drop support for some older architectures? 32-bit support was dropped. In 18.04 `fprintf` contains `MOVAPS`:

```
(gdb) disas fprintf
Dump of assembler code for function __fprintf:
   0x00007fec7a18bd80 <+0>:	sub    $0xd8,%rsp
   0x00007fec7a18bd87 <+7>:	test   %al,%al
   0x00007fec7a18bd89 <+9>:	mov    %rdx,0x30(%rsp)
   0x00007fec7a18bd8e <+14>:	mov    %rcx,0x38(%rsp)
   0x00007fec7a18bd93 <+19>:	mov    %r8,0x40(%rsp)
   0x00007fec7a18bd98 <+24>:	mov    %r9,0x48(%rsp)
   0x00007fec7a18bd9d <+29>:	je     0x7fec7a18bdd6 <__fprintf+86>
   0x00007fec7a18bd9f <+31>:	movaps %xmm0,0x50(%rsp)
...
```

`%rax` contains the number of floating point numbers passed when a function is called. And `fprintf` is called from `m_flog`, which is the function from the 3rd break point hit above where the stack address was not properly aligned. So I changed the `fprintf` in `m_flog` by adding a floating point number at the end as a parameter, to check if even under 18.04 we could get a crash with an unaligned stack.

Changed
```c
fprintf(fd,"%s.%03ld %s: ",buf,(long)spec.tv_nsec/1000000,module;
```
to
```c
fprintf(fd,"%s.%03ld %s: ",buf,(long)spec.tv_nsec/1000000,module,3.145);
```

Recompile it, run it and voilà: 
```
# ./stable/dynamips -r 1024 -X -P 1700 /images/<some 1700 .image>
Cisco Router Simulation Platform (version 0.2.21-amd64/Linux stable)
...
Segmentation fault (core dumped)
```

We get the SIGSEGV on 18.04, too. So, **the stack was misaligned already in 18.04 and we were just lucky that back then libc did not contain a `MOVAPS` in any path we took while we worked with a misaligned stack.** This was verified by adding a floating point argument to `fprintf` which causes `MOVAPS` to be executed even in older libc.

Now, how can we fix this?


### Fixing the Bug
The stack is not aligned. How can we align it?

We are using `gcc`, so `man gcc` and `/align`. Maybe something interesting pops up. We get
```
-mstackrealign
-mpreferred-stack-boundary=num
-mincoming-stack-boundary=num
```

First attempt: Compile with `-mstackrealign`.

```
... CFLAGS=-mstackrealign cmake ..
... make
... dynamips ...
```

No segfault anymore. That's good. The description in the man page did not really satisfy me. 

> On the x86, the -mstackrealign option generates an alternate prologue and epilogue that realigns the run-time stack if necessary.

What does "if necessary" mean? I guess they can only mean "unless known to be unecessary" which is not the same. Their description sounds like they knew all cases in which it is necessary. But I believe they only know some cases in which it is unnecessary, and in such cases they can leave the realignment code away to save cycles. And that, in addition to the existance of the option, indicates, that the cycles are important. That option is not for free. They also write

> See also the attribute "force_align_arg_pointer", applicable to individual functions.

So I rather add that attribute to only the functions that need rather than slowing down the whole code. But that comes at the risk of missing a function. A trade off and `dynamips` is not used to run a pacemaker. So it is not that problematic if it crashes. The missing attribute can be added then.

And I did add the attribute, recompiled, and it still works. Tested with both 1700 and 3725 to have both PPC and MIPS.

I also tried another approach: I added instructions around call sites that realign the stack. So a bad caller (i.e. the JIT generated code) would fix the stack itself before calling a function which assumes 16 byte alignment. But thinking about it caused me to decide that that is not a good solution. The reason is that in x64 some arguments are passed through registers. But at a certain number of arguments all further ones go over the stack. Called functions assume that stack parameters start at `8(%esp)`, because that is where the last PUSH before the CALL put the last/first parameter. By religning the stack just before the CALL I would have violated the assumption about `8(%esp)` referring to the the first parameter. A realignment could make the first parameter start at `16(%esp)`. To make a caller align the stack properly, you need to align it before pushing the parameters. And I did not want to look through the code for if and where parameters might be passed via the stack. So I chose the `force_align_arg_pointer` approach, letting the compiler do the work in the called functions at the cost of some unnecessary instructions being executed when a good caller calls them which would not need forced realignment.

DresslerFrank commented:
Here is an example for the effect of `force_align_arg_pointer`: https://godbolt.org/z/7EsPdao7a

```c
struct a { char x; struct c { __int128 d[2]; } e; };
void f(struct c);

void g1(struct a i) { f(i.e); }

__attribute__((force_align_arg_pointer))
void g2(struct a i) { f(i.e); }
```

turns into

```asm
g1:                                     # @g1
        sub     rsp, 40
        movaps  xmm0, xmmword ptr [rsp + 64]
        movaps  xmm1, xmmword ptr [rsp + 80]
        movaps  xmmword ptr [rsp + 16], xmm1
        movaps  xmmword ptr [rsp], xmm0
        call    f@PLT
        add     rsp, 40
        ret
g2:                                     # @g2
        push    rbp
        mov     rbp, rsp
        and     rsp, -16
        sub     rsp, 32
        movaps  xmm0, xmmword ptr [rbp + 32]
        movaps  xmm1, xmmword ptr [rbp + 48]
        movaps  xmmword ptr [rsp + 16], xmm1
        movaps  xmmword ptr [rsp], xmm0
        call    f@PLT
        mov     rsp, rbp
        pop     rbp
        ret
```
using clang.

It is not important to understand the C code. I was just trying around until clang generated something that contains `MOVAPS`. The important thing is that `g1` and `g2` are the same, except that `g2` has the attribute. And we can see that in the machine code. Because `clang` cannot exclude that `g2` is ever called with a misaligned stack, it added `and     rsp, -16` (`-16` is `0xFFF...FF0`) to clear the lower 4 bits, i.e. to make RSP contain a multiple of 16 as required by the ABI.

spikefishjohn commented:
@DresslerFrank Your reply is so interesting, thank you so much for taking the time to do that

I got as far as figuring out how to build a debug build, getting the source for libc loaded, setting break points and then even removing m_flog calls only to have it seg fault on something super basic after that. I don't recall, but I want to say GDB was telling me a number compare caused a seg fault. At that point I figured something was happening before the tzset calls that had put things in a bad way but didn't have enough understanding to figure out next where.

Is addr_perm.sh something you cooked up for this or something laying around in google somewhere? 

https://xkcd.com/386/

Nice, found that from looking up Peter. I don't know if that applies to you as well but thank you again!

DresslerFrank commented:
@spikefishjohn Yes, I created the `addr_perm.sh` just for the post, because I could not find a way to get the page permissions from GDB. But I do not like that script as it mixes two tools (`bash` and `gawk`) and contains too many lines for a simple task.
